ceph.spec.in

```diff
@@ -189,6 +189,11 @@ BuildRequires:	fuse-devel
 BuildRequires:	gcc-toolset-10-gcc-c++ >= 10.3.1-1.2
 %else
 BuildRequires:	gcc-c++
+%if 0%{?sle_version}
+# In addition to gcc-c++, we also need to explicitly require
```

Comment:
if we explicitly use g++/gcc-11, why not just use gcc11-c++ in the place of gcc-c++. i have the proposed change in the C++20 PR. see https://github.com/ceph/ceph/pull/45133/commits/b8f20444c68ec1e9be1c6a3c0b8e9be80e22c46d

Reply:
It's to work around a gcc packaging subtlety on openSUSE Leap that messes up boost's toolset detection (note that boost's toolset detection relies on "g++" being in the path somewhere).

Hopefully the following makes sense:

Normally, if we just require gcc-c++, we get the gcc-c++ package, which contains only /usr/bin/c++ and /usr/bin/g++ symlinks. gcc-c++ in turn Requires gcc$VER-c++, where $VER is whatever the default version for that distro is.

On openSUSE Tumbleweed, the default is gcc 11, so BuildRequires gcc-c++ automatically installs gcc-c++ and gcc11-c++ and we've got everything we need: /usr/bin/g++ from gcc-c++, which is a symlink to /usr/bin/g++-11 from gcc11-c++.

On openSUSE Leap, the default is gcc 7, which is obviously not new enough to build ceph. If we _only_ require gcc11-c++ on openSUSE Leap, we get  /usr/bin/g++-11, which is fine to build ceph itself, provided we specify `-DCMAKE_C_COMPILER=gcc-11 -DCMAKE_CXX_COMPILER=g++-11` when calling cmake. But, we don't have /usr/bin/g++, which means when boost's boostrap.sh runs, it can't find "g++", so it doesn't know what toolset it's using, and the build eventually fails with something like this:

```
[  520s] make[2]: *** [CMakeFiles/Boost.dir/build.make:86: boost/src/Boost-stamp/Boost-build] Error 1
[  520s] make[1]: *** [CMakeFiles/Makefile2:1929: CMakeFiles/Boost.dir/all] Error 2
```

By requiring both gcc-c++ and gcc11-c++ we ensure that there *is* a g++ binary installed so boost's toolset detection works. The fact that gcc-c++ in this case links to g++-7 doesn't matter because that's already overridden for the ceph build with -DMAKE_CXX_COMPILER.

Having written all that, I suspect an alternative fix for the boost problem might be something like this:

```diff --git a/cmake/modules/BuildBoost.cmake b/cmake/modules/BuildBoost.cmake
index 4a1aee88d0a..408b653e426 100644
--- a/cmake/modules/BuildBoost.cmake
+++ b/cmake/modules/BuildBoost.cmake
@@ -83,7 +83,8 @@ function(do_build_boost root_dir version)
   # build b2 and prepare the project-config.jam for boost
   set(configure_command
     ./bootstrap.sh --prefix=<INSTALL_DIR>
-    --with-libraries=${boost_with_libs})
+    --with-libraries=${boost_with_libs}
+    --with-toolset=gcc)
 
   set(b2 ./b2)
   if(BOOST_J)
```

...although that might break the build for FreeBSD, which AIUI uses clang not gcc.

Reply:
...oh wait, I've just realised I _can_ fix this in cmake/modules/BuildBoost.cmake - I just need to shuffle the `set(toolset)` stuff around a bit. Stay tuned...

Reply:
...yep, that worked (see 8ab5d7e)